### PR TITLE
refactor(core/biblio): export as a class

### PR DESCRIPTION
**Pull request by saschanaz**
Part of #2749.

### CODE REVIEW

File: src/core/biblio.js [hunk 1]
```diff
@@ -137,19 +95,78 @@ export async function run(conf) {
...
-  const externalRefs = split.noData.map(item => item.id);
-  if (externalRefs.length) {
-    // Going to the network for refs we don't have
-    const data = await updateFromNetwork(externalRefs, { forceUpdate: true });
-    Object.assign(biblio, data);
+
+  return idbRefs;
+}
+
+export default class {
```
Comment (saschanaz) [inline src/core/biblio.js]: Maybe we could add a proper name and check that instead. `class` is a keyword so that can't be used, do you have a good idea?

Comment (sidvishnoi) [reply src/core/biblio.js]: `klass` is pretty common

Comment (saschanaz) [reply src/core/biblio.js]: Decided to use `Class` as using uppercase matches our code style 👀

Comment (sidvishnoi) [reply src/core/biblio.js]: Smart

Comment (sidvishnoi) [reply src/core/biblio.js]: Wait. How about `Plugin`?

Comment (saschanaz) [reply src/core/biblio.js]: Every modules are Plugin. I'm not sure they are really plugins, though. (No real plugin API, need to modify profile whenever we need third party modules.)

File: src/core/base-runner.js [hunk 2]
```diff
@@ -51,13 +54,17 @@ function toRunnable(plug) {
   };
 }
 
+function isRunnableModule(plug) {
+  return plug && (plug.run || plug.exportsClass);
```
Comment (sidvishnoi) [inline src/core/base-runner.js]:
```suggestion
  return plug && (plug.run || plug.Class);
```

File: src/core/biblio.js [hunk 3]
```diff
@@ -137,19 +93,78 @@ export async function run(conf) {
...
+  /** @param {Conf} conf */
+  constructor(conf) {
+    this.conf = conf;
+  }
+
+  /**
+   * Normative references take precedence over informative ones,
+   * so any duplicates ones are removed from the informative set.
+   */
+  normalizeReferences() {
```
Comment (saschanaz) [inline src/core/biblio.js]: A separate thought: I don't quite like how this affects other modules... 🤔

Comment (sidvishnoi) [reply src/core/biblio.js]: Another separate thought: `normativeReferences` should not exist on `conf` (keep conf for user defined cobfig only?). We should rather `export const normativeReferences = new Set()` from the module where it's created

File: src/core/biblio.js (hunk 1 above)
Comment (marcoscaceres) [reply src/core/biblio.js]: Although @saschanaz is technically correct about not really being "plugins", we've always referred to these as plugins. The Class thing looks a bit weird, tbh... so I'd prefer we just call them `Plugin` for legacy reasons and hold our noses. Naming things is hard.

File: src/core/biblio.js (hunk 3 above)
Comment (marcoscaceres) [reply src/core/biblio.js]: agree... of some kind of promise getter.

File: src/core/biblio.js (hunk 1 above)
Comment (saschanaz) [reply src/core/biblio.js]: Okay, reasonable. Actually I like the Plugin name too, maybe one day we could make them real plugins...

Comment (marcoscaceres) [reply src/core/biblio.js]: I would love that :) The ultimate plugins would all be their own git repos and we'd have a nicer plugin API. What's some nice plugins APIs you've seen?

Comment (saschanaz) [reply src/core/biblio.js]: Not sure about what's "nice", but I would mention the plugins we have in our `package.json` and the APIs they are using (from Babel, ESLint etc.).

**Pull request merged**